### PR TITLE
Add config_overrides argument to initializer of MergerConfig

### LINKED ISSUE #223: Difficulty using MergerConfig
Hey, first off, thanks for the awesome library. This project solves a huge recurring problem for me when trying to parse config files. I plan to use this in a lot of projects moving forward.

From my experience, it seems like this project is CLI-first rather than API-first. This makes the library a huge pain to integrate into other projects. For example, I'm trying to make a CLI program that merges some YAML files, overwriting only specific nodes. It seems like the only way to specify these exceptions using this project is to write out a config file with a list of rules (which in my case changes on every run). This adds unnecessary complexity as I should be able to just pass these options to the `MergerConfig` class directly.

Additionally, the YAML path rules I'm writing into the config include the `:` character, which `ConfigParse` likes to convert to `=`. Then the code in `MergerConfig` assumes that the `=` symbol is part of the value, not the key. I've solve this problem by using the following nasty hack:
```python
        old_init = configparser.ConfigParser.__init__

        def new_init(self):
            old_init(self, delimiters="=")

        configparser.ConfigParser.__init__ = new_init
```

Both of these problems can be solved by allowing rules, defaults, and key to be specified in the class constructor, and only loading from a config file in the event these values are absent.

### PR DESCRIPTION
**Pull request by leviem1**
Resolves #223 by passing configuration values through MergerConfig API instead of via configuration files.

### CODE REVIEW

File: yamlpath/merger/mergerconfig.py
Comment (wwkimball) [inline yamlpath/merger/mergerconfig.py:1]: I like [what you suggested in the Issue](https://github.com/wwkimball/yamlpath/issues/223#issuecomment-1769779556) regarding `MergerConfig(logger=logger, rules={"my.yaml.path.with:colons": "left"}, defaults={"aoh", "left"})` but with a twist.  What if the constructor simply accepted additional optional mapping arguments for rules, keys, and defaults?

It seems to me that would be the most backward-compatible, API friendly approach.

That would change the constructor signature to something more like:

`MergerConfig(self, logger: ConsolePrinter, args: Namespace, defaults: MutableMapping = {}, rules: MutableMapping = {}, keys: MutableMapping = {})`

You might even consider making Properties out of defaults, rules, and keys...

**Review by wwkimball — CHANGES_REQUESTED** [review]

File: yamlpath/merger/mergerconfig.py
Comment (leviem1) [reply yamlpath/merger/mergerconfig.py:1]: I think that works. I only stuck them into the SimpleNamespace because your style linter complained of too many arguments when I tried to implement them as optional args. I'll take a peek at doing it as you suggested, but as keyword args

Comment (leviem1) [reply yamlpath/merger/mergerconfig.py:1]: Okay, I attempted the following, but ran into some errors:
```python
    def __init__(
            self,
            logger: ConsolePrinter,
            args: Namespace,
            *,
            defaults: Optional[Dict[str, str]] = None,
            keys: Optional[Dict[str, str]] = None,
            rules: Optional[Dict[str, str]] = None,
    ) -> None:
        self.log = logger
        self.args = args
        self.config: Optional[MutableMapping] = None
        self.rules: Dict[NodeCoords, str] = {}
        self.keys: Dict[NodeCoords, str] = {}
        self.user_defaults: Optional[Dict[str, str]] = defaults
        self.user_keys: Optional[Dict[str, str]] = keys
        self.user_rules: Optional[Dict[str, str]] = rules

        self._load_config()
```

Errors:
```text
************* Module yamlpath.merger.mergerconfig
yamlpath/merger/mergerconfig.py:24:0: R0902: Too many instance attributes (8/7) (too-many-instance-attributes)
yamlpath/merger/mergerconfig.py:27:4: R0913: Too many arguments (6/5) (too-many-arguments)
```

Would you like me to use a second namespace to pass these arguments instead?

Comment (wwkimball) [reply yamlpath/merger/mergerconfig.py:1]: I thought you were going to use keyword arguments?  That should keep the linter happy.  Something like:

```python
def __init__(self, logger: ConsolePrinter, args: Namespace, **kwargs: Optional[MutableMapping]):
    defaults: MutableMapping = kwargs.pop("defaults", {})
    rules: MutableMapping = kwargs.pop("rules", {})
    keys: MutableMapping = kwargs.pop("keys", {})
    <...>
```

No?

Comment (leviem1) [reply yamlpath/merger/mergerconfig.py:1]: Haha, I thought I was being clever but I was just being dumb. I'll do that!

**Comment by leviem1** [comment]
@wwkimball Just pinging you to let you know I made the changes you requested. Do you want me to start writing tests for it?

**Comment by wwkimball** [comment]
Given that the concrete behavior of the new argument is to override whatever may be in the external configuration file -- should one have been set -- please change the new property name from `user_config` to `config_overrides`.  Then, yes; please resolve any test failures and bring code test coverage back up to 100%.

**Comment by wwkimball** [comment]
Besides that, please change the merge target to the development branch.  There is normally a formal process for updating master that is tied to releases, so merges to master aren't allowed unless they come from development.

File: yamlpath/merger/mergerconfig.py [hunk 1]
```diff
@@ -24,21 +24,28 @@
...
 
         Returns:  N/A
         """
         self.log = logger
         self.args = args
-        self.config: Union[None, configparser.ConfigParser] = None
+        self.config: Optional[configparser.ConfigParser] = None
         self.rules: Dict[NodeCoords, str] = {}
         self.keys: Dict[NodeCoords, str] = {}
+        self.config_overrides: Dict[str, Any] = kwargs
```
Comment (wwkimball) [inline yamlpath/merger/mergerconfig.py:48]: Down the road, other keyword arguments could be added, so consuming the whole of `kwargs` for this purpose is problematic.  Apart from that, `self.config_overrides` is never used outside of `self._load_config()`, which is called immediately after setting the instance variable.  To add some future-proofing to this and eliminate the otherwise unused instance variable, try this:

```python
override_keys: Dict[str, Any] = **kwargs.pop("keys", {})
override_rules: Dict[str, Any] = **kwargs.pop("rules", {})
config_overrides: Dict[str, Any] = {}

if len(override_keys):
    config_overrides["keys"] = override_keys
if len(override_rules):
    config_overrides["rules"] = override_rules

self._load_config(config_overrides)
```

or something like it.

**Review by wwkimball — CHANGES_REQUESTED** [review]
Thank you for redirecting to the development branch!  Please consider the suggestion in this review.

File: yamlpath/merger/mergerconfig.py (hunk 1 above)
Comment (leviem1) [reply yamlpath/merger/mergerconfig.py]: Implemented something similar. Let me know what you think. I wanted to avoid the pop since that's a mutable operation, not that I imagine anyone would notice.

Comment (wwkimball) [reply yamlpath/merger/mergerconfig.py]: In team programming, there is a rule, "Be internally consistent."  The `pop` method is used throughout the entire rest of the code-base.  Avoiding it in one place is inconsistent.  Besides that, there's no benefit to preserving kwargs.  It's an ephemeral object generated by Python for the specific function call.  No one anywhere is looking to get it back.

**Comment by wwkimball** [comment]
The tests all pass; please resolve the kwargs matter and we'll push this along.

File: yamlpath/merger/mergerconfig.py (hunk 1 above)
Comment (leviem1) [reply yamlpath/merger/mergerconfig.py]: I appreciate you taking the time to explain your perspective. As an experienced developer, I do have a solid understanding of coding conventions and best practices. Though I agree with and will make your requested changes, in this case, I made a deliberate choice to handle the kwargs differently for the reasons I explained previously. Please don't assume my perspective results from lack of experience - I aim to make choices in the project's best interest.

Comment (wwkimball) [reply yamlpath/merger/mergerconfig.py]: From my time in Japan, one of my favorite sayings translates to, "Even monkeys sometimes fall from trees."  In other words, even experts make mistakes from time to time.  I look forward to your updated contribution.

**Comment by leviem1** [comment]
@wwkimball

File: yamlpath/merger/mergerconfig.py [hunk 2]
```diff
@@ -334,8 +346,15 @@ def _load_config(self) -> None:
 
         if config_file:
```
[the remark this answers is not in the document]
Comment (leviem1) [reply yamlpath/merger/mergerconfig.py:347]: This is just a case of not fixing an indent after making a change. Everything should work just fine be un-indenting those lines because even if the config file doesn't exist, we still use ConfigParser, adding it as a class member only if we added the overrides to it (by using `sections()`).

Though your comment made me realize that my tests should have caught this earlier, so I checked out my tests and realized that anything specified in `self.args` has a higher priority than my changes, which I think misses the mark. I'll take a deeper look into fixing this when I have a bit more time to update everything, but I'll push some changes to the tests soon to illustrate the issue.

[the remark this answers is not in the document]
Comment (leviem1) [reply yamlpath/merger/mergerconfig.py:347]: Disregard that last part, turns out it works just fine based on the tests I added. I added a fix and some tests for the fix.

**Comment by leviem1** [comment]
@wwkimball

**Comment by leviem1** [comment]
@wwkimball Bump

**Comment by leviem1** [comment]
@wwkimball Hope you had a good holiday weekend. Just wanted to bump this one last time.

**Comment by leviem1** [comment]
Hey, no worries, I was recently unemployed for three months and now I'm in a weird limbo in a contract-to-hire position, so I definitely understand. I am currently using the workaround I mentioned in my issue #223, so I'm not blocked. Take your time, I just wanted to make sure this didn't fall off your radar. And good luck at your new position!

**Comment by leviem1** [comment]
@wwkimball Hey, just wanted to see how it's going. This isn't a bump, just want to see how you've been getting along with the new job

**Comment by wwkimball** [comment]
> @wwkimball Hey, just wanted to see how it's going. This isn't a bump, just want to see how you've been getting along with the new job

I'm on a team of one working day and night plus weekends with no overtime pay while being paid 1/4 my former salary to write an ERP-light in technologies I've had to teach myself on-the-fly (Node.js, React.js, and Next.js) because my new employer has zero budget for IT but wants big-company tooling, anyway.  It's both a thrill (I love learning) and a serious challenge (I may have bit off more than I can chew).  I mistakenly told my boss I could write a simple purchase order system in two weeks.  If I'd stuck with technologies I already knew (I originally imagined Linux-Apache-PHP-PostgreSQL) and never gave any demos where everyone wanted the software to do more and more (scope creep), sure; two weeks would have been enough.  I'm now going on three months and people are already saying I'm not justifying my already meager salary (nothing to show for the time other than demos on my laptop).

I'm desperately trying to develop a deployment pipeline (Docker-based to a tiny VM running on a tiny NAS without the luxury of Jenkins) for this project before Christmas.

In a nutshell, it's rough.  And in some ways, exciting.

**Comment by leviem1** [comment]
Yikes, sorry to hear that. It's always fun to learn and take on a new challenge, but I understand how it feels when you get stuck on a project that should've been straightforward.

I know your situation is probably very different than my experiences, but I've found that more than anything it helps to just ship a turd early on and build from there. Nobody knows what they want, but they know what they don't. Plus, they want to see that stuff is "working", even if it does so poorly.

Plus, it doesn't sound like they're paying for your best either. Sometimes you have to deliver what is asked, not what is your best.

I hope your situation improves soon and I hope you had a great christmas!

**Review by wwkimball — APPROVED** [review]

**Comment by wwkimball** [comment]
@leviem1 Sorry this took so long!  Due to a minor holdup waiting for my employer's replacement server (ran the old one until it failed... 14 years!), I found some free time today.
